File: Python/Tree/M-1161. Maximum Level Sum of a Binary Tree.py
import sys


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxLevelSum(self, root: Optional[TreeNode]) -> int:
        max_lvl_sum = -sys.maxsize
        max_sum_lvl = -1

        def bfs(lvl, nodes: List[TreeNode]):
            if len(nodes) == 0:
                return

            nonlocal max_lvl_sum
            nonlocal max_sum_lvl
            lvl_sum = 0
            next_nodes: List[TreeNode] = []
            while len(nodes):
                node = nodes.pop()
                lvl_sum += node.val
                if node.left:
                    next_nodes.append(node.left)
                if node.right:
                    next_nodes.append(node.right)

            if lvl_sum > max_lvl_sum:
                max_lvl_sum = lvl_sum
                max_sum_lvl = lvl

            bfs(lvl + 1, next_nodes)

        bfs(1, [root])
        return max_sum_lvl

File: Tree/M-1161. Maximum Level Sum of a Binary Tree/M-1161. Maximum Level Sum of a Binary Tree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function maxLevelSum(root: TreeNode): number {
  let maxSum = Number.MIN_SAFE_INTEGER;
  let maxSumLvl: number = -1;

  const bfs = (lvl: number, nodes: TreeNode[]) => {
    if (!nodes.length) return;

    const nextNodes: TreeNode[] = [];
    let sum = 0;
    while (nodes.length) {
      const node = nodes.pop();
      if (node?.val) sum += node.val;
      if (node?.left) nextNodes.push(node.left);
      if (node?.right) nextNodes.push(node.right);
    }
    if (sum > maxSum) {
      maxSum = sum;
      maxSumLvl = lvl;
    }
    bfs(lvl + 1, nextNodes);
  };

  bfs(1, [root]);
  return maxSumLvl;
}
